pkg/controller/channel/channel.go

```diff
@@ -0,0 +1,155 @@
+		GVK:        prototype.GroupVersionKind(),
+		PluralName: apiv1.ChannelResourceName,
+		Status:     true,
+		Schema:     schema,
+		Categories: []string{"all", "k3os", "upgrade"},
+		ShortNames: []string{"ch", "chs", "chan", "chans"},
+		//Columns: []v1beta1.CustomResourceColumnDefinition{
+		//	{
+		//		Name:     "Latest",
+		//		Type:     "string",
+		//		JSONPath: ".status.latestVersion",
+		//		Priority: 10,
+		//	},
+		//	{
+		//		Name:     "URL",
+		//		Type:     "string",
+		//		JSONPath: ".spec.url",
+		//		Priority: 100,
+		//	},
+		//},
+	})
+	return nil
+}
+
+type Options struct {
+	ControllerName    string
+	ControllerApply   apply.Apply
+	ControllerFactory *ctl.Factory
+	PollingInterval   time.Duration
+}
+
+// RegisterHandlers registers Channel handlers
+func RegisterHandlers(ctx context.Context, options Options) {
+	channels := options.ControllerFactory.K3os().V1().Channel()
+	upgradeSets := options.ControllerFactory.K3os().V1().UpgradeSet()
+
+	ctlv1.RegisterChannelStatusHandler(ctx, channels, apiv1.ChannelLatestResolved, options.ControllerName,
+		statusHandler(ctx, options),
+	)
+	ctlv1.RegisterChannelGeneratingHandler(ctx, channels, options.ControllerApply.WithCacheTypes(upgradeSets), "", options.ControllerName,
+		generatingHandler(ctx, options), &generic.GeneratingHandlerOptions{},
+	)
+}
+
+func generatingHandler(ctx context.Context, options Options) ctlv1.ChannelGeneratingHandler {
+	timeout := upgradeset.DefaultDrainTimeout
+	return func(obj *apiv1.Channel, status apiv1.ChannelStatus) (objects []runtime.Object, _ apiv1.ChannelStatus, _ error) {
+		if obj.Status.LatestVersion == "" {
+			return objects, status, fmt.Errorf("unresolved channel")
+		}
+		objects = []runtime.Object{apiv1.NewUpgradeSet(obj.Namespace, "upgrade-"+obj.Name, apiv1.UpgradeSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: labels.Set{
+					api.LabelUpgradeChannel: obj.Name,
+				},
+			},
+			Spec: apiv1.UpgradeSetSpec{
+				Version:     status.LatestVersion,
+				Concurrency: upgradeset.DefaultConcurrency,
+				Drain: &apiv1.DrainSpec{
+					Force:            true,
+					Timeout:          &timeout,
+					IgnoreDaemonSets: true,
+					DeleteLocalData:  true,
+				},
+			},
+		})}
+		logrus.Debugf("%#v", objects)
+		return objects, status, nil
+	}
+}
+
+func statusHandler(ctx context.Context, options Options) ctlv1.ChannelStatusHandler {
+	httpClient := &http.Client{
+		CheckRedirect: func(*http.Request, []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	channels := options.ControllerFactory.K3os().V1().Channel()
+	return func(obj *apiv1.Channel, status apiv1.ChannelStatus) (apiv1.ChannelStatus, error) {
+		if obj.Spec.URL == "" {
+			return status, fmt.Errorf("missing url")
+		}
+		logrus.Debugf("Resolving %v", obj.Spec.URL)
+
+		request, err := http.NewRequestWithContext(ctx, http.MethodGet, obj.Spec.URL, nil)
+		if err != nil {
+			return status, err
+		}
+		sysid := obj.GetClusterName()
```

Comment:
This field has an unknown future in k8s.  I'd avoid using it and just generate a unique ID and store it in the configmap on startup.

Reply:
I wasn't sure where to go with either of these so I just threw something in as a placeholder until we could review.

Reply:
I found https://hub.docker.com/repository/docker/rancher/system-upgrade-controller/tags?page=1 for the next iteration. It's a nice hack.